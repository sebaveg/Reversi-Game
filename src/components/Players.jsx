import React from 'react';
import { connect } from 'react-redux';
import { changeTurn } from '../actions';

import '../assets/styles/Players.css';

const Players = (props) => {
  const handleClick = () => props.changeTurn();

  const {
    name, color, total, currentPlayer,
  } = props;
  return (
    <div className="wrapperPlayers">
      <h2>{name}</h2>
      <div className={`circle${color}`} />
      <h3>Total Disks</h3>
      <div>{total}</div>
      {color === currentPlayer
        ? (
          <>
            <h3 className="light"> Your turn </h3>
            <button type="button" onClick={handleClick}>Skip</button>
          </>
        ) : null}
    </div>
  );
};

const mapStateToProps = (state) => ({
  currentPlayer: state.players.currentPlayer,
});

const mapDispatchTopProps = {
  changeTurn,
};

export default connect(mapStateToProps, mapDispatchTopProps)(Players);
