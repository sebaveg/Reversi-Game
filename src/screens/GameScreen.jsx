import React, { Component } from 'react';
// import { ActionCreators as UndoActionCreators } from 'redux-undo';
import { connect } from 'react-redux';
import { setStarted, setWinner } from '../actions';

import Players from '../components/Players';
import Board from '../containers/Board';

import '../assets/styles/Game.css';

class GameScreen extends Component {
  constructor(props) {
    super(props);
    this.rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
    this.cols = ['1', '2', '3', '4', '5', '6', '7', '8'];
    props.setStarted(); // dispatch action
  }

  componentDidUpdate() {
    if (this.props.allowedCells === 0) {
      this.winner();
      this.props.history.push('/end');
    }
  }

  winner() {
    // This can do better
    if (this.props.playerOne.totalDisks > this.props.playerTwo.totalDisks) {
      this.props.setWinner(this.props.playerOne.name);
    }
    if (this.props.playerOne.totalDisks < this.props.playerTwo.totalDisks) {
      this.props.setWinner(this.props.playerTwo.name);
    }
    if (this.props.playerOne.totalDisks === this.props.playerTwo.totalDisks) {
      this.props.setWinner('Draw');
    }
  }

  render() {
    const {
      playerOne, playerTwo, error, historyDisks,
    } = this.props;
    return (
      <>
        {error ? <p className="error">{error}</p> : null}
        <section className="flexContainer">
          <div className="flexItem">
            <Players
              name={playerOne.name}
              color={playerOne.colorDisk}
              total={playerOne.totalDisks}
            />
          </div>
          <div className="flexItem">
            <Board />
          </div>
          <div className="flexItem">
            <Players
              name={playerTwo.name}
              color={playerTwo.colorDisk}
              total={playerTwo.totalDisks}
            />
          </div>
        </section>

        <div className="wrapper-history">

          <table cellPadding="5">
            {/* <h2 className="title">WHITE history</h2> */}
            <thead>
              <tr key="0">
                <th>Nª</th>
                <th>Movement</th>
                <th>Come back</th>
              </tr>
            </thead>
            <tbody>
              {historyDisks.posDisksWhite.map((mov, i) => (
                <tr key={`${mov[0]}${mov[1]}`}>
                  <td>{i}</td>
                  <td>{`${this.cols[mov[1]]}${this.rows[mov[0]]}`}</td>
                  <td><button type="button">Come back here</button></td>
                </tr>
              ))}
            </tbody>
          </table>

          <table cellPadding="5">
            {/* <h2 className="title">BLACK history</h2> */}
            <thead>
              <tr key="0">
                <th>Nª</th>
                <th>Movement</th>
                <th>Come back</th>
              </tr>
            </thead>
            <tbody>
              {historyDisks.posDisksBlack.map((mov, i) => (
                <tr key={`${mov[0]}${mov[1]}`}>
                  <td>{i}</td>
                  <td>{`${this.cols[mov[1]]}${this.rows[mov[0]]}`}</td>
                  <td><button type="button">Come back here</button></td>
                </tr>
              ))}
            </tbody>
          </table>

        </div>
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  error: state.game.error,
  allowedCells: state.game.allowedCells,
  historyDisks: state.disks.present,
  playerOne: state.players.playerOne,
  playerTwo: state.players.playerTwo,
});

const mapDispatchToProps = {
  setStarted,
  setWinner,
  // onJump: () => UndoActionCreators.jump(),
};

export default connect(mapStateToProps, mapDispatchToProps)(GameScreen);
