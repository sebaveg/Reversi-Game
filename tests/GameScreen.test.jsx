/* eslint-disable no-undef */
import React from 'react';
import configureMockStore from 'redux-mock-store';
import GameScreen from '../src/screens/GameScreen';

import Players from '../src/components/Players';

const initialState = {
  game: {
    error: '',
    allowedCells: '',
  },
  disks: {
    past: [],
    present: {
      posDisksWhite: [
        [3, 3],
        [4, 4],
      ],
      posDisksBlack: [
        [3, 4],
        [4, 4],
      ],
      posDisks: [],
    },
    future: [],
  },
  players: {
    playerOne: '',
    playerTwo: '',
  },
};

const mockStore = configureMockStore();

describe('<GameScreen /> Container', () => {
  let wrapper;
  beforeEach(() => {
    const store = mockStore(initialState);
    wrapper = shallow(<GameScreen store={store} />).childAt(0).dive();
  });

  it(' +++ capturing snapshot of GameScreen', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render one <section>', () => {
    const component = wrapper.find('section');
    expect(component.length).toBe(1);
  });

  it('should render one <div>', () => {
    const component = wrapper.find('div');
    expect(component.length).toBe(4);
  });

  it('should render one <Players /> components', () => {
    const component = wrapper.find(Players);
    expect(component.length).toBe(2);
  });

  it('should render one <h2>', () => {
    const component = wrapper.find('h2');
    expect(component.length).toBe(1);
  });

  it('should render one <table>', () => {
    const component = wrapper.find('table');
    expect(component.length).toBe(1);
  });
});
