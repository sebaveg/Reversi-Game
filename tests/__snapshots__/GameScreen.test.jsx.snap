// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameScreen /> Container  +++ capturing snapshot of GameScreen 1`] = `
<Fragment>
  <section
    className="flexContainer"
  >
    <div
      className="flexItem"
    >
      <Connect(Players) />
    </div>
    <div
      className="flexItem"
    >
      <Connect(Board) />
    </div>
    <div
      className="flexItem"
    >
      <Connect(Players) />
    </div>
  </section>
  <h2
    className="title"
  >
    MOVIMENTS
  </h2>
  <div
    className="wrapper-history"
  >
    <table
      cellPadding="5"
      className="table-history"
    >
      <thead>
        <tr
          key="0"
        >
          <th>
            NÂª
          </th>
          <th>
            Color Disk
          </th>
          <th>
            Position
          </th>
          <th>
            Come back
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</Fragment>
`;
